Support mapping `quadratic_field` elements to GAP and back

It would be great if we could support mapping elements of quadratic number fields to/from GAP. But the following right now gives an error:
```julia
F, sqrt3 = quadratic_field(3);
sigma = matrix(F, [1 0 0 ; 0 -1 0 ; 0 0 -1])
tau = matrix(F, [1//2 -sqrt3//2 0 ; sqrt3//2 1//2 0 ; 0 0 1])
G = matrix_group(sigma, tau)
order(G)
```
I guess we can just ask GAP to what element we should map the square root:
```
gap> Sqrt(3);
-E(12)^7+E(12)^11
```

Indeed, for now I can use this workaround:
```
F, z12 = CyclotomicField(12)
sqrt3 = sqrt(F(3))
```

So this works now, great! But it seems e.g. `natural_character` does not. Its docstring indicates that it only works over cyclotomics. Can we support this input, too? Right now I see this:
```
julia> F, sqrt3 = quadratic_field(3);

julia> sigma = matrix(F, [1 0 0 ; 0 -1 0 ; 0 0 -1]);

julia> tau = matrix(F, [1//2 -sqrt3//2 0 ; sqrt3//2 1//2 0 ; 0 0 1]);

julia> G = matrix_group(sigma, tau);

julia> natural_character(G)
ERROR: Error thrown by GAP: Error, no method found! For debugging hints type ?Recovery from NoMethodFound
Error, no 3rd choice method found for `ConjugacyClasses' on 1 arguments at ...
```
I assume this is the same problem as in #2050. So let's compute the `order` to force a natural monomorphism, and then try again:
```
julia> order(G)
12

julia> natural_character(G)
ERROR: AssertionError: fl
Stacktrace:
 [1] (::QQAbField{AnticNumberField})(a::nf_elem)
   @ Oscar.GModuleFromGap ~/Projekte/OSCAR/Oscar.jl/experimental/GModule/GModule.jl:957
 [2] #1636
   @ ./none:0 [inlined]
 [3] iterate
   @ ./generator.jl:47 [inlined]
 [4] collect(itr::Base.Generator{Vector{GroupConjClass{MatrixGroup{nf_elem, AbstractAlgebra.Generic.MatSpaceElem{nf_elem}}, MatrixGroupElem{nf_elem, AbstractAlgebra.Generic.MatSpaceElem{nf_elem}}}}, Oscar.var"#1636#1637"{QQAbField{AnticNumberField}}})
   @ Base ./array.jl:782
 [5] natural_character(G::MatrixGroup{nf_elem, AbstractAlgebra.Generic.MatSpaceElem{nf_elem}})
   @ Oscar ~/Projekte/OSCAR/Oscar.jl/src/Groups/group_characters.jl:1122
 [6] top-level scope
   @ REPL[37]:1
```

The question behind the `AssertionError: fl` is how natural a natural character is.
The square root of 3 is in a cyclotomic field, and we need an embedding into `QQAbField` in order to write down the character. Currently such a natural embedding is provided only for `nf_elem`s whose parents are of `Hecke.is_cyclotomic_type`, i.e., which have been created as cyclotomic fields.
We can define and implement such embeddings also for quadratic fields. For general abelian number fields, it will not be straightforward.

--- a/src/GAP/iso_gap_oscar.jl
+++ b/src/GAP/iso_gap_oscar.jl
@@ -41,10 +41,12 @@ function _iso_gap_oscar(F::GAP.GapObj)
        elseif GAPWrap.IsCyclotomicCollection(F)
          if GAPWrap.IsCyclotomicField(F)
            return _iso_gap_oscar_field_cyclotomic(F)
-         elseif GAPWrap.IsNumberField(F)
-           return _iso_gap_oscar_number_field(F)
          elseif F === GAP.Globals.Cyclotomics
            return _iso_gap_oscar_abelian_closure(F)
+         elseif GAPWrap.DegreeOverPrimeField(F) == 2
+           return _iso_gap_oscar_field_quadratic(F)
+         elseif GAPWrap.IsNumberField(F)
+           return _iso_gap_oscar_number_field(F)
          end
        elseif GAPWrap.IsAlgebraicExtension(F)
          return _iso_gap_oscar_number_field(F)
@@ -100,6 +102,36 @@ end
 function _iso_gap_oscar_field_cyclotomic(FG::GAP.GapObj)
    FO = cyclotomic_field(GAPWrap.Conductor(FG))[1]
    finv, f = _iso_oscar_gap_field_cyclotomic_functions(FO, FG)
+
+   return MapFromFunc(f, finv, FG, FO)
+end
+
+function _iso_gap_oscar_field_quadratic(FG::GAP.GapObj)
+   if GAPWrap.IsCyclotomicCollection(FG)
+     # Determine the root that is contained in `FG`.
+     N = GAPWrap.Conductor(FG)
+     if mod(N, 4) == 3
+       N = -N
+     elseif mod(N, 8) == 4
+       N = div(N, 4)
+       if mod(N, 4 ) == 1
+         N = -N
+       end
+     elseif mod(N, 8) == 0
+       N = div(N, 4)
+       FGgens = GAP.Globals.GeneratorsOfField(FG)
+       if any(x -> GAP.Globals.GaloisCyc(x,-1) != x, FGgens)
+         N = -N
+       end
+     end
+     FO = quadratic_field(N)[1]
+     finv, f = _iso_oscar_gap_field_quadratic_functions(FO, FG)
+   elseif GAPWrap.IsAlgebraicExtension(FG)
+     # For the moment, do not try to interpret an algebraic extension.
+     return _iso_gap_oscar_number_field(FG)
+   else
+     error("do not know how to handle FG")
+   end
 
    return MapFromFunc(f, finv, FG, FO)
 end

--- a/src/GAP/iso_oscar_gap.jl
+++ b/src/GAP/iso_oscar_gap.jl
@@ -248,11 +248,42 @@ function _iso_oscar_gap_field_cyclotomic_functions(FO::AnticNumberField, FG::GAP
    return (f, finv)
 end
 
+# Assume that `FO` and `FG` are quadratic fields with the same square root
+# in Oscar and GAP, respectively.
+# (Quadratic fields are easier to handle than general number fields.)
+function _iso_oscar_gap_field_quadratic_functions(FO::AnticNumberField, FG::GAP.GapObj)
+   flag, N = Hecke.is_quadratic_type(FO)
+   @assert flag
+
+   oO = one(FO)
+   zO = gen(FO)
+   oG = 1
+   zG = GAPWrap.Sqrt(GAP.Obj(N))
+   B = GAPWrap.BasisNC(FG, GapObj([oG, zG]))
+
+   f = function(x::Nemo.nf_elem)
+      return GAP.Obj(coeff(x,0)) * oG + GAP.Obj(coeff(x,1)) * zG
+   end
+
+   finv = function(x::GAP.Obj)
+      GAPWrap.IsCyc(x) || error("$x is not a GAP cyclotomic")
+      coeffs = GAPWrap.Coefficients(B, x)
+      coeffs === GAP.Globals.fail && throw(ArgumentError("$x is not an element oof $FG"))
+      return QQFieldElem(coeffs[1]) * oO + QQFieldElem(coeffs[2]) * zO
+   end
+
+   return (f, finv)
+end
+
 function _iso_oscar_gap(FO::AnticNumberField)
-   flag, N = Hecke.is_cyclotomic_type(FO)
-   if flag
-     FG = GAPWrap.CF(GAP.Obj(N))
+   flag1, N1 = Hecke.is_cyclotomic_type(FO)
+   flag2, N2 = Hecke.is_quadratic_type(FO)
+   if flag1
+     FG = GAPWrap.CF(GAP.Obj(N1))
      f, finv = _iso_oscar_gap_field_cyclotomic_functions(FO, FG)
+   elseif flag2
+     FG = GAPWrap.Field(GAPWrap.Sqrt(GAP.Obj(N2)))
+     f, finv = _iso_oscar_gap_field_quadratic_functions(FO, FG)
    else
      polFO = FO.pol
      N = degree(polFO)
@@ -265,7 +296,7 @@ function _iso_oscar_gap(FO::AnticNumberField)
 
      f = function(x::Nemo.nf_elem)
         coeffs = GAP.GapObj(coefficients(x), recursive = true)::GapObj
-        return GAPWrap.ObjByExtRep(fam, coeffs)
+        return GAPWrap.AlgExtElm(fam, coeffs)
      end
 
      finv = function(x::GapObj)

--- a/src/GAP/wrappers.jl
+++ b/src/GAP/wrappers.jl
@@ -5,18 +5,20 @@
 # underlying GAP function object directly and also has information about the
 # return type.
 #
-# Note that the macro GAP.@wrap has a similar purpose as @gappatribute
-# have, but works on a much lower level on purpose. We may actually phase out
-# use of @gappatribute in the future.
+# Note that the macro GAP.@wrap has a similar purpose as @gapattribute has,
+# but works on a much lower level on purpose. We may actually phase out
+# use of @gapattribute in the future.
 module GAPWrap
 
 using GAP
 
+GAP.@wrap AlgExtElm(x::GapObj, y::GAP.Obj)::GapObj
 GAP.@wrap AlgebraicExtension(x::GapObj, y::GapObj)::GapObj
 GAP.@wrap AsList(x::GapObj)::GapObj
 GAP.@wrap AsSet(x::GapObj)::GapObj
 GAP.@wrap Basis(x::GapObj)::GapObj
 GAP.@wrap Basis(x::GapObj, y::GapObj)::GapObj
+GAP.@wrap BasisNC(x::GapObj, y::GapObj)::GapObj
 GAP.@wrap CF(x::Any, y::Any)::GapObj
 GAP.@wrap CF(x::Any)::GapObj
 GAP.@wrap CHAR_FFE_DEFAULT(x::Any)::GapInt
@@ -39,6 +41,7 @@ GAP.@wrap ElementsFamily(x::GapObj)::GapObj
 GAP.@wrap ExtRepOfObj(x::GapObj)::GapObj
 GAP.@wrap ExtRepPolynomialRatFun(x::GapObj)::GapObj
 GAP.@wrap FamilyObj(x::GAP.Obj)::GapObj
+GAP.@wrap Field(x::Any)::GapObj
 GAP.@wrap FreeAbelianGroup(x::Int)::GapObj
 GAP.@wrap FreeGeneratorsOfFpGroup(x::GapObj)::GapObj
 GAP.@wrap FreeGroupOfFpGroup(x::GapObj)::GapObj
@@ -181,6 +184,7 @@ GAP.@wrap SetMaximalAbelianQuotient(x::Any, y::Any)::Nothing
 GAP.@wrap SetSize(x::Any, y::Any)::Nothing
 GAP.@wrap Size(x::Any)::GapInt
 GAP.@wrap Source(x::GapObj)::GapObj
+GAP.@wrap Sqrt(x::Int64)::GAP.Obj
 GAP.@wrap StringViewObj(x::Any)::GapObj
 GAP.@wrap UnderlyingElement(x::GapObj)::GapObj
 GAP.@wrap UnivariatePolynomialByCoefficients(x::GapObj, y::GapObj, z::Int)::GapObj

--- a/test/GAP/iso_gap_oscar.jl
+++ b/test/GAP/iso_gap_oscar.jl
@@ -133,10 +133,13 @@ end
   end
 end
 
-@testset "number fields" begin
+@testset "number fields" begin  # includes quadratic fields
    @testset "subfield $F of a cyclotomic field" for F in
      [ GAP.evalstr( "Field( [ Sqrt(5) ] )" ),
-       GAP.evalstr( "Field( [ EC(19) ] )" ) ]
+       GAP.evalstr( "Field( [ Sqrt(-1) ] )" ),
+       GAP.evalstr( "Field( [ Sqrt(6) ] )" ),
+       GAP.evalstr( "Field( [ Sqrt(-6) ] )" ),
+       GAP.evalstr( "Field( [ EC(19) ] )" ) ]  # not a quadratic field
 
      x = GAP.Globals.GeneratorsOfField(F)[1]
      y = GAP.Globals.One(F)

--- a/test/GAP/iso_oscar_gap.jl
+++ b/test/GAP/iso_oscar_gap.jl
@@ -180,6 +180,29 @@ end
    @test phi(-a^3 + a^2 + 1) == GAP.evalstr("-E(5)^2-E(5)^3")
 end
 
+@testset "quadratic number fields" begin
+   # for computing random elements of the fields in question
+   my_rand_bits(F::QQField, b::Int) = rand_bits(F, b)
+   my_rand_bits(F::AnticNumberField, b::Int) = F([rand_bits(QQ, b) for i in 1:degree(F)])
+
+   @testset for N in [ 5, -3, 12, -8 ]
+      F, z = quadratic_field(N)
+      f = Oscar.iso_oscar_gap(F)
+      @test f === Oscar.iso_oscar_gap(F)  # test that everything gets cached
+      for i in 1:10
+         a = my_rand_bits(F, 5)
+         for j in 1:10
+            b = my_rand_bits(F, 5)
+            @test f(a*b) == f(a)*f(b)
+            @test f(a - b) == f(a) - f(b)
+         end
+      end
+
+      m = matrix([z z; z z])
+      @test map_entries(inv(f), map_entries(f, m)) == m
+   end
+end
+
 @testset "number fields" begin
    # for computing random elements of the fields in question
    my_rand_bits(F::QQField, b::Int) = rand_bits(F, b)

